File: backend/app/schemas.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class DebtorBase(BaseModel):
    loan_number: str
    borrower_name: str
    ssn: str | None = None
    address: str | None = None
    phone: str | None = None
    email: str | None = None

    lender: str | None = None
    investor: str | None = None
    lien_position: str | None = None
    loan_type: str | None = None

    origination_date: date | None = None
    principal: float = 0
    rate_pct: float = 0
    term_months: int = 360
    day_count_convention: int = 360

    current_balance: float = 0
    property_value: float | None = None

    escrow_tax_monthly: float = 0
    escrow_ins_monthly: float = 0
    escrow_pmi_monthly: float = 0
    escrow_other_monthly: float = 0

    advances: float = 0
    nsf_fees: float = 0
    legal_fees: float = 0
    other_fees: float = 0
    suspense_balance: float = 0

    missed_payments: int = 0
    days_past_due: int = 0


class DebtorCreate(DebtorBase):
    pass


class DebtorUpdate(BaseModel):
    """All fields optional -- PATCH-style partial update."""
    model_config = ConfigDict(extra="forbid")

    borrower_name: str | None = None
    ssn: str | None = None
    address: str | None = None
    phone: str | None = None
    email: str | None = None
    lender: str | None = None
    investor: str | None = None
    lien_position: str | None = None
    loan_type: str | None = None
    origination_date: date | None = None
    principal: float | None = None
    rate_pct: float | None = None
    term_months: int | None = None
    day_count_convention: int | None = None
    current_balance: float | None = None
    property_value: float | None = None
    escrow_tax_monthly: float | None = None
    escrow_ins_monthly: float | None = None
    escrow_pmi_monthly: float | None = None
    escrow_other_monthly: float | None = None
    advances: float | None = None
    nsf_fees: float | None = None
    legal_fees: float | None = None
    other_fees: float | None = None
    suspense_balance: float | None = None
    missed_payments: int | None = None
    days_past_due: int | None = None


class DebtorOut(DebtorBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime


class PaymentCreate(BaseModel):
    due_date: date
    received_date: date | None = None
    amount_due: float = 0
    amount_received: float = 0
    late_charge: float = 0
    status: str = "on_time"  # on_time|late|missed


class PaymentOut(PaymentCreate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    debtor_id: int


class ReinstatementInput(BaseModel):
    """Extra one-off costs/credits that aren't part of the stored debtor
    record but are needed to price a specific reinstatement quote."""
    per_diem_days: int = 0
    fc_costs: float = 0
    title_fees: float = 0
    property_preservation: float = 0
    bankruptcy_costs: float = 0
    escrow_shortage: float = 0
    suspense_credit: float = 0
    partial_payment_credit: float = 0
    last_paid_installment: date | None = None
    due_date: date | None = None


class LoanCalculationOut(BaseModel):
    base_payment: float
    interest: float
    principal_portion: float
    per_diem: float
    per_diem_accrual: float
    late_charge: float
    total_fees: float
    total_pi: float
    escrow_monthly: float
    total_piti: float
    ltv_pct: float | None
    equity: float | None
    status: str


class ReinstatementLineItemOut(BaseModel):
    priority: int
    category: str
    description: str
    quantity: int | str
    unit_amount: float
    total: float


class ReinstatementOut(BaseModel):
    auto_missed_payments: int
    gross_reinstatement: float
    credits: float
    net_reinstatement: float
    line_items: list[ReinstatementLineItemOut]


class AmortizationRowOut(BaseModel):
    month: int
    payment_date: date | None
    payment: float
    principal: float
    interest: float
    escrow: float
    balance: float
    cumulative_interest: float
    cumulative_principal: float


class CalculationRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    debtor_id: int
    computed_at: datetime
    base_payment: float
    escrow_monthly: float
    total_piti: float
    per_diem: float
    status: str
    reinstatement_net: float | None

File: backend/src/schemas.js
const { z } = require("zod");

const dateOnly = z
  .string()
  .regex(/^\d{4}-\d{2}-\d{2}$/, "expected YYYY-MM-DD");

const debtorBaseShape = {
  loanNumber: z.string().min(1),
  borrowerName: z.string().min(1),
  ssn: z.string().nullish(),
  address: z.string().nullish(),
  phone: z.string().nullish(),
  email: z.string().nullish(),

  lender: z.string().nullish(),
  investor: z.string().nullish(),
  lienPosition: z.string().nullish(),
  loanType: z.string().nullish(),

  originationDate: dateOnly.nullish(),
  principal: z.number().default(0),
  ratePct: z.number().default(0),
  termMonths: z.number().int().default(360),
  dayCountConvention: z.number().int().default(360),

  currentBalance: z.number().default(0),
  propertyValue: z.number().nullish(),

  escrowTaxMonthly: z.number().default(0),
  escrowInsMonthly: z.number().default(0),
  escrowPmiMonthly: z.number().default(0),
  escrowOtherMonthly: z.number().default(0),

  advances: z.number().default(0),
  nsfFees: z.number().default(0),
  legalFees: z.number().default(0),
  otherFees: z.number().default(0),
  suspenseBalance: z.number().default(0),

  missedPayments: z.number().int().default(0),
  daysPastDue: z.number().int().default(0),
};

const DebtorCreate = z.object(debtorBaseShape).strict();

const DebtorUpdate = z
  .object(debtorBaseShape)
  .partial()
  .omit({ loanNumber: true })
  .strict();

const PaymentCreate = z
  .object({
    dueDate: dateOnly,
    receivedDate: dateOnly.nullish(),
    amountDue: z.number().default(0),
    amountReceived: z.number().default(0),
    lateCharge: z.number().default(0),
    status: z.enum(["on_time", "late", "missed"]).default("on_time"),
  })
  .strict();

const ReinstatementInput = z
  .object({
    perDiemDays: z.number().int().default(0),
    fcCosts: z.number().default(0),
    titleFees: z.number().default(0),
    propertyPreservation: z.number().default(0),
    bankruptcyCosts: z.number().default(0),
    escrowShortage: z.number().default(0),
    suspenseCredit: z.number().default(0),
    partialPaymentCredit: z.number().default(0),
    lastPaidInstallment: dateOnly.nullish(),
    dueDate: dateOnly.nullish(),
  })
  .strict();

module.exports = {
  DebtorCreate,
  DebtorUpdate,
  PaymentCreate,
  ReinstatementInput,
};
